fhir-validation/src/main/java/com/ibm/watson/health/fhir/validation/util/ConstraintGenerator.java

```diff
@@ -0,0 +1,182 @@
+            }
+            
+            node.elementDefinitions.add(elementDefinition);
+        }
+        
+        Tree tree = new Tree();
+        tree.root = root;
+        tree.nodeMap = nodeMap;
+        
+        return tree;
+    }
+
+    private String transform(Node node) {
+        StringBuilder sb = new StringBuilder();
+        
+        Integer min = getMin(node);
+        String max = getMax(node);
+        
+        sb.append(node.label);
+        
+        Element fixed = getFixed(node);
+        Element pattern = getPattern(node);
+        if (fixed != null) {
+            sb.append(" = ");
+            if (fixed.is(Code.class)) {
+                sb.append("'").append(fixed.as(Code.class).getValue()).append("'");
+            } else if (fixed.is(Uri.class)) {
+                sb.append("'").append(fixed.as(Uri.class).getValue()).append("'");
+            }
+        } else if (pattern != null) {
+            // TODO: pattern
+
+        }
+        
+        // TODO: value set binding
+        
+        // TODO: reference types
+        
+        if (!node.children.isEmpty()) {
+            sb.append(".where(");
+            
+            StringJoiner joiner = new StringJoiner(" and ");
+            for (Node child : node.children) {
+                joiner.add(transform(child));
+            }
+            sb.append(joiner.toString());
+            
+            sb.append(")");
+        }
+        
+        if (min != null && max != null && fixed == null) {
+            if (min == 1 && "*".equals(max)) {
+                sb.append(".exists()");
+            } else if (min == 1 && "1".equals(max)) {
+                sb.append(".count() = 1");
+            }
+        }
+        
+        return sb.toString();
+    }
+
+    private Integer getMin(Node node) {
+        for (ElementDefinition elementDefinition : node.elementDefinitions) {
+            if (elementDefinition.getMin() != null) {
+                return elementDefinition.getMin().getValue();
+            }
+        }
+        return null;
+    }
+
+    private String getMax(Node node) {
+        for (ElementDefinition elementDefinition : node.elementDefinitions) {
+            if (elementDefinition.getMax() != null) {
+                return elementDefinition.getMax().getValue();
+            }
+        }
+        return null;
+    }
+    
+    private Element getFixed(Node node) {
+        for (ElementDefinition elementDefinition : node.elementDefinitions) {
+            if (elementDefinition.getFixed() != null) {
+                return elementDefinition.getFixed();
+            }
+        }
+        return null;
+    }
+    
+    private Element getPattern(Node node) {
+        for (ElementDefinition elementDefinition : node.elementDefinitions) {
+            if (elementDefinition.getPattern() != null) {
+                return elementDefinition.getPattern();
+            }
+        }
+        return null;
+    }
+
+    static class Tree {
```

Comment:
dumb question:  why static?   honest question; I just don't think i've seen static member classes used much.

Reply:
Not sure how else to do it. We use static nested class definitions everywhere. Per https://docs.oracle.com/javase/tutorial/java/javaOO/nested.html if we used non-starting nested class definitions we’d have to instatiate them using this wonky syntax:

OuterClass.InnerClass innerObject = outerObject.new InnerClass();